uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php

```diff
@@ -3264,4 +3265,469 @@ public function testLeaveRequestBeCreatedWhenLeaveRequestHasNoWorkingDayAndValid
+    //It will now create a balance change of -4 as opposed to -3 previously.
+    $params['id'] = $leaveRequest->id;
+    $params['from_date'] = CRM_Utils_Date::processDate('2016-11-14');
+    $params['to_date'] = CRM_Utils_Date::processDate('2016-11-17');
+
+    //Since the dates as changed, the request is treated as a fresh request as if it is just being requested
+    //with entitlement balance change being same as it was before it was requested.
+    $this->setExpectedException(
+      'CRM_HRLeaveAndAbsences_Exception_InvalidLeaveRequestException',
+      'There are only ' . $entitlementBalance .' days leave available. This request cannot be made or approved'
+    );
+    LeaveRequest::create($params);
+  }
+
+  public function testUpdatingAlreadyApprovedLeaveThrowsExceptionWhenLeaveDatesNotInSamePeriodAsPreviouslyApprovedLeaveAndLeaveBalanceIsGreaterThanEntitlementBalanceForThatPeriod() {
+    $period1 = AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('2016-01-01'),
+      'end_date' => CRM_Utils_Date::processDate('2016-01-30'),
+    ]);
+
+    $period2 = AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('2016-02-01'),
+      'end_date' => CRM_Utils_Date::processDate('2016-02-28'),
+    ]);
+
+    $contactId = 1;
+    WorkPatternFabricator::fabricateWithA40HourWorkWeek(['is_default' => 1]);
+
+    HRJobContractFabricator::fabricate(
+      ['contact_id' => $contactId],
+      ['period_start_date' => '2016-01-01']
+    );
+
+    $periodEntitlement1 = LeavePeriodEntitlementFabricator::fabricate([
+      'type_id' => $this->absenceType->id,
+      'contact_id' => $contactId,
+      'period_id' => $period1->id
+    ]);
+
+    $periodEntitlement2 = LeavePeriodEntitlementFabricator::fabricate([
+      'type_id' => $this->absenceType->id,
+      'contact_id' => $contactId,
+      'period_id' => $period2->id
+    ]);
+
+    $entitlement1Balance = 3;
+    $entitlement2Balance = 0;
+    $this->createLeaveBalanceChange($periodEntitlement1->id, $entitlement1Balance);
+    $this->createLeaveBalanceChange($periodEntitlement2->id, $entitlement2Balance);
+
+    $params = [
+      'type_id' => $this->absenceType->id,
+      'contact_id' => $contactId,
+      'status_id' => 1,
+      'from_date' => CRM_Utils_Date::processDate('2016-01-05'),
+      'from_date_type' => 1,
+      'to_date' => CRM_Utils_Date::processDate('2016-01-07'),
+      'to_date_type' => 1,
+      'request_type' => LeaveRequest::REQUEST_TYPE_LEAVE
+    ];
+
+    $leaveRequest = LeaveRequestFabricator::fabricate($params, true);
+
+    $this->assertNotNull($leaveRequest->id);
+    //Entitlement balance is Zero. The three leave days have been deducted.
+    $this->assertEquals(0, $periodEntitlement1->getBalance());
+
+    //Update Leave Request dates with dates being in the second period.
+    //The balance deducted when leave request was made will not be added to the entitlement balance
```

Comment:
Since we have two entitlements in this test, it's better to clarify to which of them you referring to in this sentence